libp2p/protocols/mix/utils.nim

```diff
@@ -0,0 +1,147 @@
+import results, strutils
+import stew/base58
+import ./config
+
+const peerIdByteLen = 39
+
+proc bytesToUInt16*(data: openArray[byte]): Result[uint16, string] =
+  if len(data) != 2:
+    return err("Data must be exactly 2 bytes long to convert to uint16")
+  return ok(uint16(data[0]) shl 8 or uint16(data[1]))
+
+proc uint16ToBytes*(value: uint16): seq[byte] =
+  return @[byte(value shr 8), byte(value and 0xFF)]
+
+proc bytesToUInt32*(data: openArray[byte]): Result[uint32, string] =
+  if len(data) != 4:
+    return err("Data must be exactly 4 bytes long to convert to uint32")
+  return ok(
+    uint32(data[0]) shl 24 or uint32(data[1]) shl 16 or uint32(data[2]) shl 8 or
+      uint32(data[3])
+  )
+
+proc uint32ToBytes*(value: uint32): seq[byte] =
+  return
+    @[
+      byte(value shr 24),
+      byte(value shr 16 and 0xFF),
+      byte(value shr 8 and 0xFF),
+      byte(value and 0xFF),
+    ]
+
+proc extractPeerId(parts: seq[string], index: int): Result[seq[byte], string] =
+  if parts[index - 1] != "p2p":
+    return err("Invalid peerId component")
+  let peerIdBase58 = parts[index]
+  if peerIdBase58.len != 53:
+    return err("Peer ID must be exactly 53 characters")
+  try:
+    let peerIdBytes = Base58.decode(peerIdBase58)
+    if peerIdBytes.len != peerIdByteLen:
+      return err("Peer ID must be exactly " & $peerIdByteLen & " bytes")
+    return ok(peerIdBytes)
+  except Base58Error:
+    return err("Invalid Peer ID")
+
+proc multiAddrToBytes*(multiAddr: string): Result[seq[byte], string] =
```

Comment:
is this something that can live with multiaddress type?

Reply:
The serialization used here differs from that used in multiaddress. This one is particular for mix defined in https://github.com/vacp2p/rfc-index/blob/cc8c7a0ac923d382e5ca778ffa44dbc21c80ec33/vac/raw/mix.md#84-address-and-delay-encoding